### PR TITLE
Basement node

### PR DESCRIPTION
**Pull request by etwest**
Originally we were planning on introducing basement nodes but I realized it makes just as much sense (if not more) to just make the leaves smaller.

This is because with the introduction of the circular queue we have a nice way of handling the fact that many updates will go to the same node.

Additionally, as we have more and more nodes in the graph, the size of the sketch will approach (but likely not exceed) 1 MB. This means the case where we have to do many leaf flushes for one parent flush will be very unlikely.

### CODE REVIEW

File: src/buffer_tree.cpp [hunk 1]
```diff
@@ -47,13 +48,16 @@ nodes, int workers, bool reset=false) : dir(dir), M(size), B(b), N(nodes) {
 	buffer_size     = M; // probably figure out a better solution than this
 	max_buffer_size = 2 * M;
 	backing_EOF     = 0;
+	leaf_size       = floor(24 * pow(log(N) / log(2), 3)); // size of leaf proportional to size of sketch
```
Comment (Victor-C-Zhang) [inline src/buffer_tree.cpp:51]: why not use function `log2`?

File: src/buffer_tree.cpp [hunk 2]
```diff
@@ -47,13 +48,16 @@ nodes, int workers, bool reset=false) : dir(dir), M(size), B(b), N(nodes) {
...
+	leaf_size       = (leaf_size < page_size)? page_size : leaf_size; //enforce size of at least page_size
 
 	// malloc the memory for the root node
-	root_node = (char *) malloc(max_buffer_size);
+	root_node = (char *) malloc(buffer_size);
 	root_position = 0;
 
 	// malloc the memory to use for reading buffers
 	read_buffer = (char *) malloc(max_buffer_size);	
+	backup_read_buffer = (char *) malloc(2 * leaf_size);
```
Comment (Victor-C-Zhang) [inline src/buffer_tree.cpp:60]: why is this a backup?

File: src/buffer_tree.cpp (hunk 1 above)
Comment (etwest) [reply src/buffer_tree.cpp:51]: No good reason. I'll change it:)

File: src/buffer_tree.cpp (hunk 2 above)
Comment (etwest) [reply src/buffer_tree.cpp:60]: Because it's used when we need to flush a leaf while still in the process of flushing the parent

**Review by Victor-C-Zhang — APPROVED** [review]
LGTM

**Pull request merged**